### PR TITLE
Add gammapy.astro.darkmatter

### PR DESCRIPTION
**Pull request by joleroi**
This PR introduces ``gammapy.astro.darkmatter`` with some utilities for dark matter analysis. At the moment it is a collection of things I have left over from my thesis. Define the scope of this package and its relation to existing packages is yet to be done.

### CODE REVIEW

**Comment by cdeil** [comment]
@joleroi - Please always put a PR description, even for WIP PRs, especially if it's a major addition like here: 500+ lines, probably 1000 soon with tests and docs.

Generally I'm Ok to add such things to `gammapy.astro` for now, as long as it's clear that this is prototyping generally the future of `gammapy.astro` is a bit uncertain. E.g. it could happen that we split it off at some point and keep Gammapy focused on analysis.
@joleroi - Agreed?
Could you please update  http://docs.gammapy.org/dev/astro/index.html a bit here to describe it that way?

Another thing that would be very useful to have here would be a paragraph or bullet list in the docs what other codes have on this topic. (It's always OK to read their docs, just not their code if they aren't license compatible.)

Looks like in Fermi ST they have a `DMFitFunction` (see [here](https://fermi.gsfc.nasa.gov/ssc/data/analysis/scitools/source_models.html), and I think in Fermi-LAT the SED profiles in Fermipy and this is heavily used for DM studies: [dmsky](https://github.com/fermiPy/dmsky) and [dmpipe](https://github.com/fermiPy/dmpipe) . Giacomo has something [here](https://github.com/giacomov/astromodels/blob/master/astromodels/functions/dark_matter/dm_models.py), and Nathan put something in Gammalib (see [here](https://cta-redmine.irap.omp.eu/issues/1520)).

From a very quick look at https://ascl.net/code/search/dark%20matter I see https://ascl.net/1201.012 and https://ascl.net/1110.002 , and from https://www.astropy.org/affiliated/ I see  http://gala.adrian.pw/ and https://galpy.readthedocs.io/, [halotools](http://halotools.readthedocs.io/en/latest/source_notes/empirical_models/phase_space_models/nfw_profile_source_notes.html) and  with Google I found https://bitbucket.org/weniger/gamlike  and https://github.com/erccarls/GammaLike

@joleroi - Could you please write the docs page for `gammapy.astro.darkmatter` first and summarise what functionality you want to put there (doesn't all have to be in this PR, I mean the vision what you want it to be say in a few month or a year), and how it compares to what has been or is being used in the field e.g. by Fermi-LAT, HESS and in CTA and by other existing codes.

Also: do you want to get a review here, i.e. should we find a DM expert now, or would you prefer to just merge this soon with only superficial code review, and then have the conversation / review later when this is merged and the docs are online? I think either way is fine to proceed, but if you want a review of the functionality / equations here, we could probably find someone next week at the CTA meeting.

File: gammapy/astro/darkmatter/profiles.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    Parameters
+    ----------
+    r_s : `~astropy.units.Quantity`
+        Scale radius, :math:`r_s`
+    rho_s : `~astropy.units.Quantity`
+        Characteristic density, :math:`\rho_s`
+
+    References
+    ----------
+    * `arXiv:astro-ph/9611107 <https://arxiv.org/abs/astro-ph/9611107>`_
```
Comment (cdeil) [inline gammapy/astro/darkmatter/profiles.py]: Prefer the final journal reference over arxiv references, no?
So http://adsabs.harvard.edu/abs/1997ApJ...490..493N

File: gammapy/astro/darkmatter/profiles.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    * `arXiv:1012.4515 <https://arxiv.org/abs/1012.451>`_
```
Comment (cdeil) [inline gammapy/astro/darkmatter/profiles.py]: "Paper identifier '1012.451' not recognized"
I think it's this one: http://adsabs.harvard.edu/abs/2011JCAP...03..051C

File: gammapy/astro/darkmatter/spectra.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    Based on the precompute models by `Cirelli et al.
+    <http://www.marcocirelli.net/PPPC4DMID.html>`_. All available
+    annihilation channels can be found there. The dark matter mass will be set
+    to the nearest available value. The spectra will be available as
+    `~gammapy.spectrum.models.TableModel` for a chosen dark matter mass and
+    annihilation channel.
+
+    References
+    ----------
+    * `arXiv:1012.4515 [hep-ph] <https://arxiv.org/abs/1012.4515>`_
```
Comment (cdeil) [inline gammapy/astro/darkmatter/spectra.py]: http://adsabs.harvard.edu/abs/2011JCAP...03..051C

File: gammapy/astro/darkmatter/utils.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        TODO: Needs to be implemented more efficiently
+        """
+        angular_dist = make_separation_map(
+            geom=self.map_.geom,
+            position=self.center)
+        rmin = angular_dist.quantity.to('rad').data * self.distance
+        rmax = self.distance
+        val = [self.profile.integral(_, rmax) for _ in rmin.flatten()]
+        jfact = u.Quantity(val).to('GeV2 cm-5').reshape(rmin.shape)
+        self.diff_jfact = WcsNDMap(data=jfact.value,
```
Comment (cdeil) [inline gammapy/astro/darkmatter/utils.py]: Is it possible / desirable to avoid coupling this DM code to our maps classes?
Shouldn't the models mostly provide evaluate and then map filling is done elsewhere?
Seems your  coupling evaluate and map making here in `compute_differential_jfactor` in a non-modular way (e.g. doesn't allow ot use with HEALPix map).

File: gammapy/astro/darkmatter/utils.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    prim_flux : `~gammapy.astro.darkmatter.PrimaryFlux`
+        Primary gamma-ray flux
+    x_section : `~astropy.units.Quantity`
+        Velocity averaged annihilation cross section, $\langle \sigma\nu\rangle$
+    energy_range : tuple of `~astropy.units.Quantity`
+        Energy range for the map
+
+    References
+    ----------
+    * `arXiv:1012.4515 [hep-ph] <https://arxiv.org/abs/1012.451>`_
```
Comment (cdeil) [inline gammapy/astro/darkmatter/utils.py]: Same broken reference to https://arxiv.org/abs/1012.451

File: gammapy/spectrum/utils.py [hunk 6]
```diff
@@ -199,12 +199,12 @@ def apply_edisp(self):
...
     """
-    Integrate 1d function using the log-log trapezoidal rule. If scalar values
-
-    for xmin and xmax are passed an oversampled grid is generated using the
-    ``ndecade`` keyword argument. If xmin and xmax arrays are passed, no
-    oversampling is performed and the integral is computed in the provided
-    grid.
+    Integrate 1d function using the log-log trapezoidal rule.
+    
+    If scalar values for xmin and xmax are passed an oversampled grid is
```
Comment (cdeil) [inline gammapy/spectrum/utils.py]: This DM PR needs tests and docs and probably a bit of review. Please split out this unrelated cleanup edit in `gammapy/spectrum/utils.py`, Ok to do such small cleanup commits directly in `master`.

**Comment by joleroi** [comment]
@cdeil @wood thanks for you input. Obviously I was a little naive when starting this project before really looking into what already exisits. I will spent some time to check the packages you mentioned (especially ``DMFitFunction``), document what they do and see if anything needs to be added to gammapy at all.

**Comment by cdeil** [comment]
> see if anything needs to be added to gammapy at all.

In any case you should add the docs page with an overview of what exists and how to do something.
I'm also not against adding code, as long as you see it as a starting point that might be replaced or rewritten or removed as we integrate with other astrophysics and analysis packages.

**Comment by joleroi** [comment]
> as long as you see it as a starting point that might be replaced or rewritten or removed as we integrate with other astrophysics and analysis packages.

Doesn't this hold for almost every LOC in gammapy :laughing: ?

**Comment by cdeil** [comment]
@joleroi - Yes, I just wanted to caution given that potentially you're putting a lot of effort on this now.

**Comment by joleroi** [comment]
I added a brief overview of other packages. I still don't know if what I put here is useful for anyone. I will try to find someone from the HESS astroparticle working group to review this.

**Comment by cdeil** [comment]
@joleroi Thanks!

It's pretty hard to get people to review or even try out PRs. You could also just merge this, so that it shows up in the online docs and is easier to try out, and get some feedback that way in the coming weeks.
